keylime/allowlistdbproxy.py

```diff
@@ -0,0 +1,247 @@
+
+
+class AllowlistEntry:
+    """AllowlistEntry represents a named allowlist in the DB"""
+
+    session: Session
+
+    update_lock: threading.Lock
+    last_update_check: float
+    update_timeout: float
+
+    allowlist_checksum: str
+    allowlist_id: int
+    verifier_allowlist_db: VerifierAllowlist
+    verifier_allowlist: VerifierAllowlist
+
+    query_checksum_stmt: Select
+
+    @staticmethod
+    def from_db(session: Session, allowlist_id: int):
+        """Create an AllowlistEntry object from an entry in the DB"""
+        try:
+            verifier_allowlist_db = session.query(VerifierAllowlist).filter_by(id=allowlist_id).one()
+        except NoResultFound:
+            return None
+        except SQLAlchemyError as e:
+            logger.error("SQLAlchemy Error: %s", e)
+            return None
+        return AllowlistEntry(session, allowlist_id, verifier_allowlist_db.checksum, verifier_allowlist_db)
+
+    def __init__(
+        self,
+        session: Session,
+        allowlist_id: int,
+        checksum: str,
+        verifier_allowlist_db: VerifierAllowlist,
+        update_timeout=0,
+    ):
+        """constructor"""
+        self.session = session
+
+        self.update_lock = threading.Lock()
+        self.last_update_check = time.monotonic()
+        self.update_timeout = update_timeout
+
+        self.allowlist_checksum = checksum
+        self.allowlist_id = allowlist_id
+        self.verifier_allowlist_db = verifier_allowlist_db
+        self.verifier_allowlist = AllowlistEntry.__copy_verifier_allowlist(self.verifier_allowlist_db)
+
+        # Use .where rather than .filter_by for older SQLAlchemy
+        self.query_checksum_stmt = select(VerifierAllowlist.checksum).where(VerifierAllowlist.id == self.allowlist_id)
+
+    @staticmethod
+    def __copy_verifier_allowlist(tocopy: VerifierAllowlist) -> VerifierAllowlist:
+        """Make a copy of the verifier_allowlist_db"""
+        # This is the object we will give to the user; it's basically a read-only object without
+        # access to the DB and should prevent that the user of the object sees changing fields
+        # while we refresh() the object from the DB
+        verifier_allowlist = VerifierAllowlist()
+        verifier_allowlist.id = tocopy.id
+        verifier_allowlist.name = tocopy.name
+        verifier_allowlist.ima_policy = tocopy.ima_policy
+        verifier_allowlist.tpm_policy = tocopy.tpm_policy
+        verifier_allowlist.checksum = tocopy.checksum
+
+        return verifier_allowlist
+
+    def __has_checksum_changed(self) -> bool:
+        """Check whether the checksum of the allowlist in the DB has changed"""
+        rnd = int.from_bytes(random.randbytes(4), byteorder="little")
+
+        verifier_allowlist = self.session.execute(
+            f"SELECT checksum,{rnd} FROM allowlists WHERE id={self.allowlist_id}",
+            execution_options={"compiled_cache": None},
+        ).one()
```

Comment:
I have been entertaining the testfarm quite a bit. For some reason the test runs don't seem to always see the updated checksum in the `-c update` test. I don't see this problem locally but I don't run the same type of database setup locally. My guess it has something to with caching but I am trying to eliminate caching. Maybe there's a path in the code that doesn't update the checksum but why would this not always happen. Odd.